core/domain/classifier_services.py

```diff
@@ -224,3 +226,117 @@ def delete_classifier(classifier_id):
     classifier_data_model = classifier_models.ClassifierDataModel.get(
         classifier_id)
     classifier_data_model.delete()
+
+
+def get_classifier_training_job_from_model(classifier_training_job_model):
+    """Gets a classifier training job domain object from a classifier
+       training job model.
+
+    Args:
+        classifier_training_job_model: Classifier training job instance in
+            datastore.
+
+    Returns:
+        classifier_training_job: Domain object for the classifier training job.
+    """
+    return classifier_domain.ClassifierTrainingJob(
+        classifier_training_job_model.id,
+        classifier_training_job_model.algorithm_id,
+        classifier_training_job_model.exp_id,
+        classifier_training_job_model.exp_version_when_created,
+        classifier_training_job_model.state_name,
+        classifier_training_job_model.status,
+        classifier_training_job_model.training_data)
+
+def get_classifier_training_job_by_id(job_id):
+    """Gets a classifier training job from a job_id.
+
+    Args:
+        job_id: str. ID of the classifier training job.
+
+    Returns:
+        classifier_training_job: Domain object for the classifier training job.
+
+    Raises:
+        Exception: Entity for class ClassifierTrainingJobModel with id not
+            found.
+    """
+    classifier_training_job_model = (
+        classifier_models.ClassifierTrainingJobModel.get(job_id))
+    classifier_training_job = get_classifier_training_job_from_model(
+        classifier_training_job_model)
+    return classifier_training_job
+
+
+def _create_classifier_training_job(classifier_training_job):
+    """Creates classifier training job model in the datastore given a
+       classifier training job domain object.
+
+    Args:
+        classifier_training_job: Domain object for the classifier training job.
+
+    Returns:
+        job_id: str. ID of the classifier training job.
+    """
+    job_id = classifier_models.ClassifierTrainingJobModel.create(
+        classifier_training_job.algorithm_id, classifier_training_job.exp_id,
+        classifier_training_job.exp_version_when_created,
```

Comment:
@prasanna08 If we were to keep the exploration_version updated, then, whenever you would save a new version of the exploration, you would have to update the exploration version for _all_ the classifiers which are associated with some state of that exploration. So, if there are 3 classifiers associated with that exploration, you'd have to retrieve 3 training jobs, 3 classifier_data instances so as to update their "exploration_version". Alternatively, you could just modify the exploration_version of the existing classifier but this would be problematic when the exploration is rolled back.

I'd also like to add to your last point - it is not really necessary to forward propagate. In the old design doc, we had mentioned this clearly - when the exploration version is updated because of a safe change (something that doesn't affect the classifier's training data), then, the classifier also won't change. So, there is no use forward propagating the classifier as we really don't want to create 10 different classifiers for 10 different exploration_versions. If there is really a big disadvantage of not keeping the exploration_version updated, then, I'd agree that keeping it updated is fine. But the only problem I can see is that we no longer would be able to retrieve the classifier for the specific version of an exploration (retrieving for the latest version would be easy: just get the classifier with the "most recent" exploration_version_when_created). However, I don't know when we'd need to find the classifier for an old exploration version other than for debugging purpose (in which case, we could create a service that would add the training job to the database using a specific exploration_version's training_data).

I remember clearly that I and @seanlip had a long talk on this topic almost a year ago and we had the same discussion with different use cases in our mind. Finally, we had decided that exp_version_when_created was simpler.